File: src/Common/TGenericStream.cpp

#include "TGenericStream.h"
#include "hxshared.h"

namespace VITALENGINE
{

//===========================================================
// TGenericStream::TMemoryStream()  
//===========================================================
TGenericStream::TGenericStream(DWORD initialSize_, DWORD increment_)
{
 readOnly=false;
 allocatedSize=initialSize_;
 initialSize=initialSize_;
 increment=increment_;
 basePointer=new BYTE[initialSize];
 curPointer=basePointer;
 length=0;
}

//===========================================================
// TGenericStream::TGenericStream()   2
//===========================================================
TGenericStream::TGenericStream(void* pointer, DWORD length, bool readOnly)
{
 basePointer=(BYTE*)pointer;
 curPointer=basePointer;
 this->length=length;
 allocatedSize=length;
 increment=0;
 this->readOnly=readOnly;
}


//===========================================================
// TGenericStream::~TGenericStream()
//===========================================================
TGenericStream::~TGenericStream()
{
 if (!readOnly) delete[] basePointer; 
}

//===========================================================
// void TGenericStream::Grow()
//===========================================================
void TGenericStream::Grow(DWORD destsize)
//grow to destsize+increment
{                                       
 BYTE* tmp;
 if (increment!=0)
  {
   tmp=new BYTE[destsize+increment];
   allocatedSize=destsize+increment;
  }
   else
  {
   tmp=new BYTE[destsize*2];
   allocatedSize=destsize*2;
  }  
  
 memcpy(tmp,basePointer,length);

 *(DWORD*)&curPointer=(DWORD)curPointer-(DWORD)basePointer+(DWORD)tmp;
 delete[] basePointer;
                                   
 basePointer=tmp;
}


//====================================================
// 
//====================================================
BYTE* __stdcall TGenericStream::GetBasePointer()
{
 return basePointer;
}

//====================================================
// 
//====================================================
BYTE* __stdcall TGenericStream::GetCurPointer()
{
 return curPointer;
}

//====================================================
// 
//====================================================
bool __stdcall TGenericStream::isReadOnly()
{
 return readOnly;
}

//====================================================
// 
//====================================================
DWORD __stdcall TGenericStream::GetLength()
{
 return length;
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::Write(const void* Data, DWORD count)
{
 assert(!readOnly);
 if (GetPos()+count>allocatedSize) Grow(GetPos()+count);
 memcpy(curPointer,Data,count);
 DWORD length1=GetPos()+count;
 if (length<length1) length=length1;
 curPointer+=count;
}

//====================================================
// 
//====================================================
DWORD __stdcall TGenericStream::Read(void* Data, DWORD count)
{
 int tmp=length-GetPos();
 if (tmp<0) return 0;  //read beyond the end

 if (count>(DWORD)tmp) count=tmp;
 memcpy(Data,curPointer,count);
 curPointer+=count;
 return count;
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::Seek(DWORD pos)
{
 *(DWORD*)&curPointer=(DWORD)basePointer+pos;
}

//====================================================
// 
//====================================================
DWORD __stdcall TGenericStream::GetPos()
{
 return (DWORD)curPointer-(DWORD)basePointer;
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::Clear()
{
 assert(!readOnly);
 delete[] basePointer;

 allocatedSize=0;
 basePointer=new BYTE[0];
 curPointer=basePointer;
 length=0;
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::FastClear()
{
 assert(!readOnly);
 curPointer=basePointer;
 length=0;
}
 
//====================================================
// 
//====================================================
void __stdcall TGenericStream::GrowToPos(int DestSize)
{
 if (DestSize==-1) DestSize=(int)GetPos();

 assert(!readOnly);
 if (allocatedSize<(DWORD)DestSize) 
  {
   Grow(DestSize);
  }
 length=DestSize;
 curPointer=basePointer;
 INCREMENT_DWORD(curPointer,length);
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::Skip(DWORD count)
{
 Seek(GetPos()+count);
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::SetLength(DWORD newLength)
{
 if (allocatedSize<(DWORD)newLength) 
  {
   Grow(newLength);
  }
 length=newLength;
}

//====================================================
// 
//====================================================
void __stdcall TGenericStream::Compact()
{
 if (allocatedSize==length) return;
 
 BYTE* tmp;
 tmp=new BYTE[length];
 allocatedSize=length;

 memcpy(tmp,basePointer,length);

 *(DWORD*)&curPointer=(DWORD)curPointer-(DWORD)basePointer+(DWORD)tmp;
 delete[] basePointer;
                                   
 basePointer=tmp;
}


} //namespace

File: Common/T_GenericStream.pas
//==============================================================================
// hxGrid framework
// Copyright (C) 2007 by Roman Lut
// hax@deep-shadows.com
// http://www.deep-shadows.com/hax/
//==============================================================================

unit T_GenericStream;

interface
uses Windows, Classes, I_GenericStream, SysUtils;

//===========================================================
// TGenericStream
//===========================================================
type TGenericStream = class (TObject, IGenericStream)
  public

   //====== begin COM interface ==========
   function QueryInterface(const IID: TGUID; out Obj): HResult; stdcall;
   function _AddRef: Integer; stdcall;
   function _Release: Integer; stdcall;

   function GetVersion(): DWORD;

   function GetBasePointer(): PByteArray; stdcall;

   function GetCurPointer(): PByteArray; stdcall;

   function isReadOnly(): boolean; stdcall;

   function GetLength(): DWORD; stdcall;

   procedure Write(var data; count: DWORD); stdcall;

   function  Read(var data; count: DWORD): DWORD; stdcall;

   procedure Seek(pos: DWORD); stdcall;

   function GetPos(): DWORD; stdcall;

   procedure Clear(); stdcall;

   procedure FastClear(); stdcall;

   procedure GrowToPos(DestSize : integer = -1); stdcall;

   procedure Skip(count: DWORD); stdcall;

   procedure SetLength(newLength: DWORD); stdcall;

   procedure Compact(); stdcall;

   //====== end COM interface ==========

   constructor Create();
   destructor Destroy(); override;

   protected
    _RefCount : integer;
    memStream: TMemoryStream;

end;


implementation

//==============================================================================
//==============================================================================
constructor TGenericStream.Create();
begin
 _RefCount:=1;
 memStream:=TMemoryStream.Create();
end;

//==============================================================================
//==============================================================================
function TGenericStream.QueryInterface(const IID: TGUID; out Obj): HResult; stdcall;
begin
 result:=0;
end;

//==============================================================================
//==============================================================================
function TGenericStream._AddRef: Integer; stdcall;
begin
 result:=InterlockedIncrement(_RefCount);
end;

//==============================================================================
//==============================================================================
function TGenericStream._Release: Integer; stdcall;
begin
 result:=InterlockedDecrement(_RefCount);
 if result=0 then free();
end;

//==============================================================================
//==============================================================================
destructor TGenericStream.Destroy();
begin
 memStream.Free();
end;

//==============================================================================
//==============================================================================
function TGenericStream.GetVersion(): DWORD;
begin
 result:=IGENERICSTREAM_VERSION;
end;

//==============================================================================
//==============================================================================
function TGenericStream.GetBasePointer(): PByteArray; stdcall;
begin
 result:=PBYTEARRAY(memStream.memory);
end;

//==============================================================================
//==============================================================================
function TGenericStream.GetCurPointer(): PByteArray; stdcall;
begin
 result:=PBYTEARRAY(memStream.memory);
 inc(DWORD(result),memStream.Position);
end;

//==============================================================================
//==============================================================================
function TGenericStream.isReadOnly(): boolean; stdcall;
begin
 result:=false;
end;

//==============================================================================
//==============================================================================
function TGenericStream.GetLength(): DWORD; stdcall;
begin
 result:=memStream.size;
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.Write(var data; count: DWORD); stdcall;
begin
 memStream.Write(data,count);
end;

//==============================================================================
//==============================================================================
function  TGenericStream.Read(var data; count: DWORD): DWORD; stdcall;
begin
 result:=memStream.Read(data,count);
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.Seek(pos: DWORD); stdcall;
begin
 memStream.Position:=pos;
end;

//==============================================================================
//==============================================================================
function TGenericStream.GetPos(): DWORD; stdcall;
begin
 result:=memStream.Position;
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.Clear(); stdcall;
begin
 memStream.Clear();
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.FastClear(); stdcall;
begin
 memStream.Clear();
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.GrowToPos(DestSize : integer = -1); stdcall;
begin
 if (DestSize=-1) then memStream.SetSize(memStream.position)
                  else memStream.SetSize(DestSize);
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.Skip(count: DWORD); stdcall;
begin
 assert(false);
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.SetLength(newLength: DWORD); stdcall;
begin
 memStream.SetSize(newLength);
end;

//==============================================================================
//==============================================================================
procedure TGenericStream.Compact(); stdcall;
var
 pushpos: DWORD;
 memStream1 : TMemoryStream;
begin
 pushPos:=memStream.Position;
 memstream1:=TMemoryStream.Create();
 memStream1.LoadFromStream(memStream);
 memStream.Destroy();
 memStream:=memStream1;
 memStream.Seek(pushPos, soFromBeginning);
end;


end.
